### ISSUE TITLE
Feature Discussion: Rutime Configuration

### ISSUE THREAD

**Issue by TAAPArthur** [nsxiv/nsxiv#108]
I had laughed at the idea of text based configuration before, but recently saw some good examples of some simple confs and thought I'd open up the discussion. I'm honestly indifferent to any of this landing in master or being patches or cited as an example of something that should never be done. Especially curious how people who don't compile from source or otherwise feel about these.

So wanted to just list what we have and what we could have. Going to expand the scope of discussion to ***configuring nsxiv and or changing its behavior after it has been installed by a package manager***

Already implemented

0. config.h (can control most things)
1. Keybindings (can only be bound to predefined functions util #50 )
2. key-handler and similar etc (doesn't have access to full state)
3. Cli args (limited support). With a wrapper can also support ENV vars
4. Xresource (limited support)
5. Accepting artifical key/mouse events (I had initially marked this as something not already done before)

With the feature matrix between config.h, cli and Xresource being very incomplete.

Possible additional ideas:

| Num | Est diff size |  Description | Pros | Cons | Patch description | Notes
| --- | ------------- | -----------| ------| ----- | --------------- | ------- | 
| 1   | 10;0  | Install a shared lib and compile your config file with it. | May not matter much to this project since the code base is small,but it is generally faster than compiling everything from scratch. Also allows all the flexibility you can get from the config file | Requires all libs needed to build which may be extra on Ubuntu or similar |  New make target; Compiling with -fpic | What xmonad does
| 2  |2;0| Like the above but instead of a shared lib, install all the source | Basically the same a compiling from source, but can track  versions with your package manager | Hacky | N/A |
| 3  |20;1| Text based config file read on startup to set config vars; sample format `VAR_NAME VALUE` | | Less versatile than config.h | Read file line-by-line, parse string find corresponding var and set arg to that variable|Part of what mpv does
| 4  |20;1 | Text based config file read on startup to set keybindings; sample format `MOD-KEY_NAME FUNC_NAME arg` | | Less versatile than config.h | Basically the same but for function names; can use X functions to convert char to KEY_SYM| Part of what mpv does
| 6 | 10-20;10 | IPC including sockets, message queues, reading from pipe/stdin/fifo, x_client_message etc | Most versatile when interacting with the running process | | Setting up the specific method, getting the message and parsing it like we did for 3/4 | 
| 7 | -10;-10 | Specify default arguments from config file | No wrapper needed; Allows every option to be controlled by the config file | May  | Stop explicitly setting all the values on `_options`. They already default to 0 and by not specifying then either a user could have a constructor to set them or we define `opt_t _options` in config.h and then users have full control over the defaults. | Would have to decide if  boolean options like `-b` toggle behave or continue to force it to a value. I guess this itself could be configurable. Such effort isn't included in the diff count.

Some notes
- May go without saying, but any of these that would modify config vars (outside of 2) would need the var to not be static/const by definition. 
- I have a prototype of extending keybindings without touching the defaults [here](https://github.com/TAAPArthur/nsxiv/commit/b66950520cad196b70eff5a4ddf9b5e514c73bcd) Basically it can be done without breaking anything but has little benefit outside of the above usecases so haven't tried to land it.
- Text configs and IPCs would need a list of config var/func name. This could easily be generated from existing code and isn't counted towards the size of the diff. 
- For "Est diff size", the first part is how large I expect the entire patch to be (not counting stuff that can be generated). The second is how little of it would be a meaningful change in an existing source file in nsxiv. So not counting config.def.h or the makefile or any new file that may exists nor any list of functions/config names that could be autogenerated.
- Nothing listed here breaks existing behavior

So with that said wanted to open discussion to see if anyone would find any of these useful. Or do we all just compile from source and not care? Please feel free to ask for clarification or implementation details if you are interested or suggest something not listed here.

Keep in mind that some of these are simpler/more versatile/compliment our existing methods of configuration. For example we don't need Xresource if we had full text based configuration for config values (after #95 at least). But on the flip side if read config options from a text file, letting users set whatever config options they wanted from Xresources would basically come free. Not saying we want to, but just an example that some of the benefits/costs of thees can be shared with across rows.

**Comment by eylles** [comment]
I think the "chapest" option that we can get more utility from in the long run is ipc, in theory we already have a form of ipc with inotify and the basic image reloading, adding a one way socket (do we get any use from a 2 way socket?) with a per instance file that is monitored by inotify and gets read and parsed every time it is modified shouldn't be too massive of a change, the problem as always comes from parsing the strings from the file into functions and arguments.

Perhaps a bit off topic but i think being able to load different keyhandlers could be useful.

**Comment by TAAPArthur** [comment]
If we used a socket we wouldn't need a inotify. We could just poll directly on it. But yeah it would be fairly straight forward and not invasive. Parsing the file isn't any harder than read a text file; given we add a array fo function names to func pointers and all arguments were either nothing or an it it'd be trivial. Even supporting strings wouldn't be bad if we marked the type in this array.

>Perhaps a bit off topic but i think being able to load different keyhandlers could be useful.

Can't that already be done? Wouldn't it just involve swapping the external file with a different one? Or having the script track state and when a certain key is pressed it swaps its list of bindings with another one and writes it down so the next instance remembers it. Can even split it up by pid of the caller.

**Comment by eylles** [comment]
i was thinking how exactly xmonad config works? i know it is a haskell source file that gets compiled and then xmonad loads/reloads it, but xmonad itself doesn't get recompiled, on the regard of the build deps (-dev packages in debian and such) the ones for nsxiv ain't nearly as invasive as requiring an installation of ghc and the many libraries that a xmonad install in debian requires so in that regard there would be no problem i think, a benefit of mimicking the way xmonad is configured would be providing users of binary distros a way to config nsxiv without having to recompile the whole program.

**Comment by N-R-K** [comment]
Imo other than IPC, I don't really see much value on the other suggestions. There are already many other image viewers that allows very versatile runtime configuration. Compiling nsxiv doesn't really require many deps nor does it take long, so I don't see much reason for wanting to avoid it.

- - -

Sometime around the first release of nsxiv, I received an email from a user who was worried if we took the "evolution" approach instead of the "suckless" approach. There are already many "evolved" image viewers, but not many suckless ones. So I'm also biased towards keeping things that way.

**Comment by eylles** [comment]
iirc specterwm also works like xmonad in that only the config needs to be recompiled bot the whole program.

I can see the benefit of it, specially if #76 or any version of it ever makes it to mainstream nsxiv then having a config that can be edited without compiling the whole program would benefit users of binary distros as they could just install nsxiv with the package manager and then have some config.c that they could add their custom functions and compile it without needing to recompile the whole program.

Does that go too much against suckless? idk i have seen both spectrwm and xmonad be considered as suckless software...

**Comment by TAAPArthur** [comment]
> Does that go too much against suckless? idk i have seen both spectrwm and xmonad be considered as suckless software...

I'm not sure how you define "suckless" but don't think everyone has the same definition. People would call mpv suckless and it has text based configuration. Same with vim which can be very bloated if you turn all compile options on. 

Personally I have a loose definition of somewhere that doesn't suck to hack on, configure nor build. This change doesn't cross that line for me. And completly agree with all the benefits listed for it


>Other than IPC, I don't really see much value on the other suggestions.

Interesting. IPC were towards the bottom in terms of practicality to me. I know they are powerful and all, but what semi-realistic usecase is there that couldn't be achieved by sending key presses?

>Sometime around the first release of nsxiv, I received an email from a user who was worried if we took the "evolution" approach instead of the "suckless" approach. There are already many "evolved" image viewers, but not many suckless ones. So I'm also biased towards keeping things that way.

Would you mind clarify this viewpoint? I can get not wanting to populate the codebase, but what's the real difference between this and the other new features we've added. From my perspective we've just been going by what we think is useful rather than an aim of pure minimalism. I'm thinking of the title patch, the new image formats, probably some other minor changes. I'm not saying those shouldn't have gone in but what's the objective difference between them? Adding an option to install it as a shared lib wouldn't qualify as an "evolution". Maybe the pure runtime configuration would. But would it because of the complexity, the size of the diff, the lack of practicality or do we want to just minimize features? I'm sure we could go to far if we aren't careful, but I question if we are currently consistent.

**Comment by eylles** [comment]
IPC does have value, at least i view it as valuable, if we ever add a funcyion to add images to a running instance of nsxiv we could simply leave it as an IPC controlled function without needing to add an additional interface like what zathura has.

**Comment by TAAPArthur** [comment]
I don't doubt the value. I just have a hard time picturing the use case. Why do you want to add an image to running instance? At worst one could just write to some predefined file, have a keybinding to to read said file line by line and add the images.  Of course such a solution may not be useful for 3rd party integration since it would require a custom config.

Couldn't nsixv also listen on a directory and couldn't then user then just symlink whatever files they wanted to add/remove there?

**Comment by N-R-K** [comment]
> Would you mind clarify this viewpoint?

Text based runtime configuration is definitely not suckless. As for mpv, it's a more minimalist media player (compared to the rest) but not strictly a "suckless style" player. And Xmonad is an exception, not really a norm.

As for accepting artificial keystrokes, that can be done via `xdotools` and thus should be considered out of the project scope. And default args can be done via wrapper quite easily as well.

**Comment by eylles** [comment]
I'm just saying if the binding of arbitrary functions gets merged only users of source distros would gain benefit from it, meanwhile a xmonad stile config will benefit binary distros and perhaps even source distro users as modifiying the config would be faster.

**Comment by TAAPArthur** [comment]
> Text based runtime configuration is definitely not suckless. As for mpv, it's a more minimalist media player (compared to the rest) but not strictly a "suckless style" player. And Xmonad is an exception, not really a norm.

What I was trying to get at as that "suckless" is a vague term and don't think we all share the same defintion nor do I think it matters. I'm betting there is some deeper reason that you dislike some of these and was hoping to learn it and the difference between that and some things that do make it into the repro. sxiv/nsxiv have many differences from surf/dwm and I don't see that as an intrinsic problem.

Before responding to this, consider the following example:

>As for accepting artificial keystrokes, that can be done via xdotools and thus should be considered out of the project scope.

Yeah xdotools works. But nsxiv doesn't accept events from XSendEvent which is what some OSK use to communicate. Since the initial post I have confirmed that. And can conveniently state it takes 5 chars to fix. So why shouldn't we make the fix? Please humour and explain without using the term "suckless" or similar.

I kinda feel like you just hear the term "runtime config" and are writing everything off. Not gonna lie, I had the same gut reaction. I think we just need concrete well defined reasons to determine what gets into master, becomes a patch, delegated to custom configs or guarded by ifdegs. I think any of these are simple enough such that an argument could be made for any of them with the main difference being convenience. I just can't justify why we are fine with adding extra img formats but right off the set of these as "not suckless".

>And default args can be done via wrapper quite easily as well.

Yes they can. But you didn't address the fact that doing it inside nsxiv would actually save lines

**Comment by TAAPArthur** [comment]
@N-R-K Just to clarify the addition of webp wouldn't fit my definition of suckless based on how large the patch is for something could be useless to a fair amount of users. Have similar thoughts to gifs more so on account of the related bugs and the desktop files/icons for only affecting specific users. But these went into master fine. And there are issues for similar things. These seem much more invasive and less useful (my opinion) that most of what's being proposed. So I don't understand your metric for writing off these suggestions. I understand that you specifically may not have approved all of these, but I'm using our existing feature set to gauge what level of changes are viable

**Comment by N-R-K** [comment]
I think you're misunderstanding, I didn't "write it off" or reject the proposal entirely. The title says "feature discussions", so I simply shared my thoughts.

As for animated webp, the rational is simple. An image viewer should be able to view images. *Support for single frame images should go into imlib2*, but since imlib2 is does not support multi-frame images we had to deal with on our end, the same way we dealt with gif.

**Comment by eylles** [comment]
On the multi frame Ytrewq13 decided to look into apng, and we may still have to look into tiff's multipage.

You can say that those are for the sake of completeness features as we should ideally support the animate version of every image format that imlib2 supports, with other image formats like heif and avif, altho there are third party loaders for those we should push for them to be added upstream.

And onto suckless software itself it ain't uncommon for suckless to break their own rules in regard of features, the best example is probably dwm, there are window managers far more suckless in every possible way than dwm, so for this project that originally was more of a suckless adjacent we can afford to stray a bit from suckless in order to improve usability.

**Comment by TAAPArthur** [comment]
> I think you're misunderstanding, I didn't "write it off" or reject the proposal entirely. The title says "feature discussions", so I simply shared my thoughts.

Glad to hear and thanks for clarifying. And I think I'm finally understanding your perspective.

>An image viewer should be able to view images.

I believe this is the answer I was looking for. I agree none of these proposal affect the ability to view images; they just add some possible conveniences. And if the mainline functionality should be limited to only patches that are either bug fixes or related to image formats, that seems reasonable. I completely disagree but I think I see where you are coming from. Please correct my if I missed the mark.

If that was the "official" stance, I'd agree that ipc/text based config should just be patches. I don't think we are all in agreement on that stance though. Like @eylles is saying, that stance could be taken to add every format directly in nsxiv (instead of say directly in imglib). And I'd rather keep the code base smaller rather than smart more than what imglib gives us by default.

In contrast my metric is more on usability/extensibility/convince against complexity/lines of code. Which is why I use (1). All the benefits of config.h while being able to receive upstream updates without risk of conflicts. And since this can be done with basically just a wrapper and a new make target I see little downside.

**Comment by N-R-K** [comment]
In order for IPC to be useful I think it'll have to be a 2 way one.

Oh, and just to be clear, I said I saw some value in IPC, but didn't say that it's going to mainline. And quite frankly, I'm not quite seeing a 2 way IPC be simple enough for mainline. But perfectly fine if someone wants to prove me wrong.

**Comment by eylles** [comment]
what would be the usecase for 2 way ipc? the most usecase i see for 2 way ipc on the current codebase is perhaps query the path of the current image, or the paths of all currently loaded images.

**Comment by N-R-K** [comment]
> query the path of the current image, or the paths of all currently loaded images.

And the only thing that cannot be done via keybindings are adding/reordering/resorting the current images. If you can think of more then do tell.

But if those are the only usecases for IPC then being able to query the current images makes perfect sense.

- - -

Also if those really are the only usecases, then maybe we should look into adding those features (either patch or mainline) rather than generic IPC.

**Comment by N-R-K** [comment]
> being able to to sort by modification date would be nice

That can be trivially done via shell wrapper. If we're talking about (re)sorting once nsxiv is open, then I can see it being useful. But didn't we reject randomize slideshow request due it being possible via shell wrapper?

> I don't think it would be much worse than 1-way

Should've wrapped (2 way) in parenthesis. If we want to go down the IPC path then it should be robust and well thought out, not half-assed. 1 or 2 way I'm not quite seeing a robust IPC implementation being "simple".

**Comment by TAAPArthur** [comment]
> That can be trivially done via shell wrapper. If we're talking about (re)sorting once nsxiv is open, then I can see it being useful. But didn't we reject randomize slideshow request due it being possible via shell wrapper?

I don't disagree and it is simple enough to implement in config.h if one felt like it. I think the difference between this and randomize-slideshow is the usefulness to do it on demand (unless one wanted to re-randomize?)

>If we want to go down the IPC path then it should be robust

Sure so let me go into more detail. Underlying assumption is that it is trivial to go from window id to pid. And am assuming each function that can be triggered can is of the same type as those for keybindings (eg accepts at most a single int and returns bool). Also assuming we have the name to function pointer array.

XClientMessage -- Find the window the window and send the special event. The event can contain 5 fields. Suppose they contain function name, arg (can be whatever if the function doesn't use it) and the senders win id. We lookup the function pass the arg and we are done for 1-way ipc. If we want to client to know the result, then we send another client message back to the client using their win id with the result of the function. Think this is the simplest.

STDIN/name pipe -- Write a message of the format func name, arg. We could either have a dedicated terminated ("\n" or \0) or just enforce it to be an exact size. Think the latter may run into an issue if someone writes a bad message, so would lean towards writing the terminated before and after the real message. On Linux at least we get 4k atomic read/write which should be sufficient to handle multiple simultaneous writers. To respond back we could have the client include their pid and just signal them depending on the result of the call. Or just put the result in /tmp/nsxiv/output_pid where the client knows to look if we wanted to support string responses. Think sockets and message queues are similar enough to this.

**Comment by TAAPArthur** [comment]
I had hinted in the initial poist how many lines I'd think each method would take. Think it should provide a rough estimate. Not sure if you didn't see/don't believe my claims or using some other metric for simplicity. I don't think LOC is the best, but don't know how else to objectively quantify it. I think the impls would be straight forward enough too

**Comment by N-R-K** [comment]
Thats just the IPC itself, if we want to

> query the path of the current image, or the paths of all currently loaded images.
> adding/reordering/resorting the current images.

Then I don't think it will be that small of diff, would it?

> I think the difference between this and randomize-slideshow is the usefulness to do it on demand (unless one wanted to re-randomize?)

Yes, once the loop is completed once it won't be random anymore. Same with sort by date, you can do it only at init, but not at runtime. I personally deal with sort (by date) via my file manager, it feeds all the pics directly to nsxiv (preserving the sort order).

**Comment by TAAPArthur** [comment]
> Then I don't think it will be that small of a diff, would it?

What would make it large? For the file based fps, we can loop and get all the paths in like half a dozen lines and just write them out. We already know what the paths are and technically already export it for the current image. For the XClientMessage... we could literally do the same thing actually. I guess this could inspire us to write a bunch of small functions like this and don't know when we'd cross the small line. On the brightside these functions could basically live in their own file.

>I personally deal with sort (by date) via my file manager, it feeds all the pics directly to nsxiv (preserving the sort order).

That's neat. When I spawn nsxiv manually I generally point to a directory and have a wrapper than calls ls on said directory to sort isn't that bad... although a tad annoying if one wanted to be robust. To address that and the lack of dynamicness, I'm thinking of 
(A) having an ability to have add cli args from config.h. Like specifying `--extra-args "some-custom arg"` and having this call some func-pointer that can be set. So with this one extra arg we can ignore any future "personal" cli requests. 
(B) and then adding the ability to restart nsxiv with exactly the same args. All this involves is making the argv array public and calling exec. With this you could detect if new images are added (which we strangely don't seem to do). And maybe supplement that with a startup hook (after cli parsing but before creating the window) to optionally run some extra func. All together this would be like half a dozen lines. Then completely in a custom config, 
1. could have a func to set env vas to persevere state like file position and sort mode, 
2. have a keybinding to set sort mode to date call restart
3. Have a startup hook that restores state based on the env vars and/or sorts the files

And if we wanted to generalize, I believe we could lump these as simple hooks. Each one is basically just a funcion pointer but gives the user power to do basically anything if we put them in the right places. Technically if we drop one in the event loop code, one could implement support for IPCs if they are clever.

**Comment by N-R-K** [comment]
I don't think we should be changing anything in mainline because of external patches, if that's what you're asking.

**Comment by TAAPArthur** [comment]
Yup that was basically what I was asking. It was more specific to simple refactoring but believes that falls under "anything". Would you care to elaborate more on your rationale? Making changes so that the codebase is easier to hack and where patches are easier to apply doesn't sound negative to me. Phrasing it as being for external patch X or just because it makes the code cleaner are just ways to think about the same end result.
